Allow managers not to expose a remote API port

aaronlehmann opened the pull request:
We'd like Docker to support swarm services without needing to run `swarm init`. One of the preconditions for this is that the swarm manager needs to operate and be useful without exposing a port to the outside world.

This PR is split into multiple parts, which will each become a separate PR for ease of review. The first few commits are focused on allowing the agent and certificate issuance/renewal to work without relying on a TCP connection to back to the same process. This works by exposing servers such as the dispatcher and node CA on the unix socket (or equivalent), and using that instead of a TCP connection. In theory we could codegen something to patch the calls straight through to the handlers (with some stuff injected onto the context to identify the caller as the local node), but for now, using the unix socket is good enough. This required some changes to the raft proxy to let it inject the local node identity onto the context when it's calling the handler locally.

- [x] Fix demotion (#1829)
- [x] Expose the dispatcher, node CA, and a few other services on the UNIX socket. Inject the local node identity when they receive connections this way. (#1828)
- [x] Add a `connectionbroker` package. This is a small abstraction on top of `Remotes` that provides a gRPC connection to a manager. If running on a manager, it uses the unix socket, otherwise it will pick a remote manager using `Remotes`. This allows things like agent and certificate renewal to work even on a single node with no TCP port bound. (#1850)
- [x] Convert the agent and CA client to use `connectionbroker`. (#1851)
- [ ] Separate binding ports from creating the manager. Add methods to `Manager` and `Node` that allow binding a port after the manager is already running. A small change to raft allows the raft code to get the machine's external address after raft is already running, instead of relying on having it already. (This PR)
- [ ] A simple integration test that binds a port after starting the first manager. (This PR)


cc @aluzzardi @LK4D4 @diogomonica @cyli @tonistiigi @stevvooe

aluzzardi commented:
/cc @ehazlett @icecrime

aluzzardi commented:
Awesome!

This also solves many issues we had with the local agent not connecting to the local manager, such as status drifting (e.g. Node both "down" and "reachable").

LK4D4 commented:
@aaronlehmann can we split it into different prs for easier reviews? Connbroker change permeates almost all files, so it's hard to find other changes.

aaronlehmann commented:
@LK4D4: This PR is split into 3 commits. I thought reviewing commit by commit would make sense, but if you think it's easier, I can split this into multiple PRs.

Would you like me to open a separate PR with just the connection broker commit, or would you like me to split that one into smaller pieces?

aaronlehmann commented:
If you'd like, I could split it into a sequence of commits and/or PRs like this:

1. Expose dispatcher/NodeCA/etc on local socket, modify proxy to support this.
2. Add connection broker package
3. Update existing code to use connection broker instead of `remotes`
4. Add late-binding support to manager and node
5. Add integration test

LK4D4 commented:
@aaronlehmann splitting plan will make it way easier for me to review. I think it's better to open different PRs.
Thanks!

aaronlehmann commented:
OK. If it's alright with you, I'll leave this one open as a complete PR to show the big picture, and also open one PR at a time to get the individual pieces merged.

ehazlett commented:
nice!  design sgtm

aaronlehmann commented:
Split the commits as discussed, and opened #1828 to review/merge the first piece. I've seen a few integration test failures with just this first one locally, but it's tricky to reproduce and I can't understand why this change would cause issues (since it's just exposing some services in a way that nothing uses yet). I wonder if the failures are related to a recent change on master instead of the changes here.

aaronlehmann commented:
I have a theory about why this PR makes the integration tests flaky. I think it changes the timing enough that a demoted node can end up with a worker certificate before Raft creates outgoing connections. Then the manager blocks in `WaitForLeader` instead of shutting down on demotion like it's supposed to.

I'm going to open a PR to change demotion so that the manager is always shut down explicitly after a role change instead of shutting itself down. Even if that doesn't fix the problem here, it's something that badly needs to be done.

aaronlehmann commented:
> I'm going to open a PR to change demotion so that the manager is always shut down explicitly after a role change instead of shutting itself down. Even if that doesn't fix the problem here, it's something that badly needs to be done.

WIP PR at #1829. Unfortunately, it doesn't work yet.

diogomonica commented:
Design LGTM.

aaronlehmann commented:
I've rebased this on top of #1829 and made a few fixes. #1829 introduced an interesting issue. With this PR, a node that's a manager will always connect to itself instead of a random other manager for dispatcher and CA operations. However, if it has been demoted, it may not be able to issue certificates, and the CA client would not retry with other managers, so the demotion would not complete. I've fixed this so that after the first attempt to renew a certificate, the client will explicitly prefer random managers over the local connection. Also, a timeout was necessary for the certificate renewal request. If the node was demoted, it can get into a state where it waits forever for a leader.

This seems quite stable now. I'm not seeing any more issues with the integration tests.

I'll keep this open as a meta-PR, and update it as individual bits get merged. #1829 should be merged before #1828, and after that I can open more PRs for the other pieces.

aaronlehmann commented:
#1829 was merged. Rebased.

aaronlehmann commented:
#1828 was merged. Rebased to remove that part of this meta-PR.

Opened #1850 as the next step in getting these pieces merged.

aaronlehmann commented:
`connectionbroker` was merged in #1850. Opened next PR: #1851.

aaronlehmann commented:
#1851, which converts code to use `connectionbroker`, was merged. Rebased this PR.

Since all the prerequisites have been merged in separate PRs, let's turn our focus back to this PR, instead of opening a new one with the same content.

dperny commented on manager/manager.go:
```diff
@@ -111,8 +111,7 @@ type Config struct {
 type Manager struct {
-	config    *Config
-	listeners []net.Listener
+	config Config
```
Why the change from pointer to value?

aaronlehmann replied:
> Why the change from pointer to value?

We now modify `m.config` in `BindRemote` and `BindControl`, and we want these modifications to be private to the manager, rather than potentially overwriting fields in a structure that was passed in by something else.

dperny commented:
Most of my review comments are not actually problems with the code, they're just a few things that I don't understand.

cyli commented on manager/manager.go:
```diff
@@ -71,7 +71,7 @@ type Config struct {
 	// RemoteAPI is a listening address for serving the remote API, and
 	// an optional advertise address.
-	RemoteAPI RemoteAddrs
+	RemoteAPI *RemoteAddrs
```
Out of curiosity why the change to a pointer?

aaronlehmann replied:
So it can be `nil` before it's initialized

cyli replied:
I meant why not just check the default value, as opposed to nil?  If `ListenAddr` is `""`, there is no remote api set, right?  I'm not arguing that's better or anything - more just asking pointers are generally preferred over checking against the default value.

aaronlehmann replied:
What you're suggesting should work fine. I felt that since `RemoteAddrs` is a struct, it might be confusing to check one particular field to see if the struct is initialized.

cyli replied:
Ah ok, thank you for explaining

cyli commented on manager/state/raft/membership/cluster.go:
```diff
@@ -236,6 +236,31 @@ func (c *Cluster) ReplaceMemberConnection(id uint64, oldConn *Member, newConn *M
+	newMember.RaftMember.Addr = newAddr
+	c.members[id] = &newMember
+
+	if oldMember.RaftMember.Addr != newAddr {
```
Non-blocking:  maybe we can move the copying of the `RaftMember` object inside this block, since we don't need to make an update if the address hasn't changed?

aaronlehmann replied:
Fixed

aaronlehmann commented on manager/state/raft/raft.go:
```diff
@@ -240,6 +243,51 @@ func NewNode(opts NodeOptions) *Node {
+		defer cancel()
+
+		isLeader := atomic.LoadUint32(&n.signalledLeadership) == 1
+		for !isLeader {
```
As we discussed offline, currently only the leader can submit proposals such as configuration changes. For now, this `SetAddr` function is intended to address the narrow use case of a single-node cluster that didn't have ports bound before. Since the cluster only has one member, this node must be the leader, but there can be delays before that leadership is confirmed, for example at startup. This loop waits until the current node becomes the leader.

If/when we extend `SetAddr` to handle more general cases, we'll need some other approach here, such as an RPC we can use to tell the leader to update a node's address.

aaronlehmann commented on manager/manager.go:
```diff
@@ -152,21 +157,102 @@ func (l *closeOnceListener) Close() error {
+		if err := m.BindRemote(context.Background(), *config.RemoteAPI); err != nil {
+			l := <-m.controlListener
+			l.Close()
+			return nil, err
```
Yes. It's safe to read from `m.controlListener` because this is before `Run`, so nothing else will be reading from that channel.

aaronlehmann commented on manager/manager.go:
```diff
@@ -132,9 +131,15 @@ type Manager struct {
 	stopped bool
+
+	remoteListener  chan net.Listener
+	controlListener chan net.Listener
```
Correct.

aaronlehmann commented on manager/manager.go:
```diff
@@ -152,21 +157,102 @@ func (l *closeOnceListener) Close() error {
+}
+
+// BindRemote binds a port for the remote API.
+func (m *Manager) BindRemote(ctx context.Context, addrs RemoteAddrs) error {
```
Order changed

aaronlehmann commented on manager/manager.go:
```diff
@@ -132,9 +131,15 @@ type Manager struct {
 
-	mu      sync.Mutex
+	mu     sync.Mutex
+	addrMu sync.Mutex
```
Added comments

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -240,6 +243,51 @@ func NewNode(opts NodeOptions) *Node {
+
+	n.opts.Addr = addr
+
+	if n.IsMember() {
```
I think it's better to just return here on `!n.IsMember()`

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -240,6 +243,51 @@ func NewNode(opts NodeOptions) *Node {
+		leadershipCh, cancel := n.SubscribeLeadership()
+		defer cancel()
+
+		isLeader := atomic.LoadUint32(&n.signalledLeadership) == 1
```
`signalledLeadership` is really weird name now :) we should call it `isLeader` and use it instead of `isLeader` function. Otherwise, it's quite confusing here.
Probably in another PR.

aaronlehmann commented:
Rebased, PTAL

aaronlehmann commented:
Rebased again, and added `t.Parallel` to the new test.

This should be ready to merge.

LK4D4 commented on manager/manager.go:
```diff
@@ -186,17 +171,88 @@ func New(config *Config) (*Manager, error) {
+	if m.config.ControlAPI != "" {
+		return errors.New("manager already has a control API address")
+	}
+	m.config.ControlAPI = addr
```
I wonder if we should clear this in case of error.

aaronlehmann replied:
Fixed

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -259,6 +262,56 @@ func (n *Node) ReportUnreachable(id uint64) {
+			if leadershipChange == IsLeader {
+				isLeader = true
+			}
+		case <-ctx.Done():
```
I'm not sure, but seems like this ctx might be `context.Background()` and it's possible that node could be stopped during this loop. Maybe we should use `WithContext`.

aaronlehmann replied:
Switched to `WithContext`.

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -1092,6 +1151,11 @@ func (n *Node) reportNewAddress(ctx context.Context, id uint64) error {
 	if err != nil {
 		return err
 	}
+	if oldAddr == "" {
```
Is this really possible?

aaronlehmann replied:
Yeah, I encountered it in the late-binding integration test. It happens when you create a manager without a bound port, then later bind one. The node update that adds the address gets committed to raft at some point, and new nodes that join the cluster may take a little while to catch up to that.

LK4D4 replied:
It's quite weird. `newPeer` calls `grpc.Dial` which is supposed to check if address is empty :/

aaronlehmann replied:
Hmm, should we avoid calling `grpc.Dial` if `newPeer` is passed an empty address?

LK4D4 replied:
I dunno, there shouldn't be empty addresses and I wonder why it works and registers in peer list at all.

LK4D4 commented on manager/manager.go:
```diff
@@ -186,17 +171,88 @@ func New(config *Config) (*Manager, error) {
+		m.config.RemoteAPI = nil
+		// The context isn't used in this case (before (*Manager).Run).
+		if err := m.BindRemote(context.Background(), *config.RemoteAPI); err != nil {
+			l := <-m.controlListener
```
This can hang forever. So probably you need to check len(l).

aaronlehmann replied:
Yes, this looks wrong. This code should only run if `config.ControlAPI != ""`. I've fixed that.

cyli replied:
Is there a use case where there is a remote API, but no control API?

aaronlehmann replied:
There isn't currently any use case without a control API. It's structured that way to be consistent with `BindRemote`.

LK4D4 commented on manager/manager.go:
```diff
@@ -186,17 +171,88 @@ func New(config *Config) (*Manager, error) {
+	if config.ControlAPI != "" {
+		m.config.ControlAPI = ""
+		if err := m.BindControl(config.ControlAPI); err != nil {
+			return nil, err
```
Shouldn't we close listener here as well?

aaronlehmann replied:
If `BindControl` returns an error, there is no listener to close.

LK4D4 commented on manager/state/raft/membership/cluster.go:
```diff
@@ -143,6 +143,7 @@ func (c *Cluster) UpdateMember(id uint64, m *api.RaftMember) error {
 		return nil
 	}
 	oldMember.RaftMember = m
+	c.broadcastUpdate()
```
nice!

LK4D4 commented on manager/state/raft/raft.go:
```diff
@@ -342,6 +398,9 @@ func (n *Node) JoinAndStart(ctx context.Context) (err error) {
 	}
 
 	// join to existing cluster
+	if n.opts.Addr == "" {
```
I don't really understand events sequence. What ensures that `BindRemote` is called before `Run`?

aaronlehmann replied:
It does not need to be called before `Run`. See the new test `TestServiceCreateLateBind`, where it is called afterwards.

LK4D4 replied:
Then I don't understand how does it work. Isn't it possible to call `JoinAndStart` before `BindRemote` and get `attempted to join raft cluster without knowing own address`?

aaronlehmann replied:
You need to bind first if you are trying to join a cluster, but if you're starting a new cluster then you hit the `n.opts.JoinAddr == ""` branch and having an address is not required.

LK4D4 commented:
LGTM
